### ISSUE TITLE
Inheriting secondary files

### ISSUE THREAD

**Issue by StarvingMarvin** [common-workflow-language/common-workflow-language#127]
If the app adds another index to the input file that already has secondary files, there is no way to say "result is all existing secondary files, plus this one that I add".

**Comment by tetron** [comment]
How would you like this to work?  A flag specifying whether it is a "merge" or "replace" operation?

**Comment by StarvingMarvin** [comment]
I guess replace as default is OK, and than some

    "inheritFrom": "input_id"

somewhere to signify merge.

**Comment by mr-c** [comment]
Suggestion: `inheritSecondaryFilesFrom`

**Comment by tetron** [comment]
Proposed design from codefest discussion:

* "secondaryFiles" is moved from "inputBinding" up to parameter definition
* If a input parameter has secondaryFiles, they must be presented to the application as if they are in the same directory (may require symlinking)
* inputBinding gets field "outdirStage" of either "link" or "copy" which means the input file & its secondaryFiles must be placed in the outputDirectory (via symlink or copy) before running
* outputBinding field "propagateFile" which links to input parameter & combines primary and secondaryFiles from source with any additional secondaryFiles that are declared.

**Comment by mr-c** [comment]
Is this critical for Draft 3 or can it wait for Draft 4?

**Comment by ntijanic** [comment]
I think it's fairly critical since chaining indexers is a common pattern (tool1 makes index1, tool2 makes index2, tool3 requires index1 and index2).

**Comment by tetron** [comment]
I don't have this implemented or even fully designed.  If this is critical for draft-3 then it's going to delay the spec.  Can we address it early in draft-4?

**Comment by ntijanic** [comment]
Yeah, don't need to delay release of draft-3 because of this, but:

- May want to include it as add-on to draft-3 if it blocks describing common pipelines or pipeline patterns.
- May want to drop `secondaryFiles` field from `inputBinding` completely since it doesn't really do anything without the accompanying requirement to stage them. Or just move it one nesting level up?

The `outdirStage` and `propagateFile` are underspecified anyway. Can continue discussion here to fully spec them.

**Comment by portah** [comment]
Can you also consider another situation, our notorious indexers, which is most likely closely connected with yours. So, particular example ```samtools faidx``` it is already in workflows repository with test files.
The main idea of this tool to make index of .fa files. When I wrapped it I used createFileRequirement and this part works. What does not work is the output.
* I can say that output is the particular index file ```glob: $(inputs.input.path.split('/').slice(-1)[0]+".fai")```, then I have troubles to pass this info within workflow because usually we use indices as secondaryFiles but this on in glob
* Or we can say the output is actually input and index file is secondaryFiles but then we have no control to actually have it 
```
[fai_build] fail to open the FASTA file dm3_chr4.fa
Final process status is success
{
    "index": {
        "path": "/Users/porter/Work/scidap/workflows/test/dm3_chr4.fa", 
        "size": 1378901, 
        "secondaryFiles": [], 
        "class": "File", 
        "checksum": "sha1$fdf5d59bd5acf52bcaa368bbf5308212f776c8c8"
    }
}%   
```

So the question is how to deal whit this confusion?

**Comment by ntijanic** [comment]
> Or we can say the output is actually input and index file is secondaryFiles but then we have no control to actually have it

This would be the preferred way. Not sure what you mean by "no control". It should be possible to specify output like:

```yaml
outputs:
  - id: indexed
    type: File
    outputBinding:
      glob: $(inputs.input.path.split('/').slice(-1)[0]  # Or just "*.fa(sta)"
      secondaryFiles: "^.fai"
```

**Comment by portah** [comment]
> This would be the preferred way. Not sure what you mean by "no control". It should be possible to specify output like:
```yaml
outputs:
  - id: indexed
    type: File
    outputBinding:
      glob: $(inputs.input.path.split('/').slice(-1)[0]  # Or just "*.fa(sta)"
```
The glob file always exist after cwl execution and if the tool is silently failed or succeed but with different file
```yaml
      secondaryFiles: "^.fai"
```
will be just an empty array. And I probably wrong but secondaryFile is not mandatory ?

**Comment by tetron** [comment]
the specific change "move `secondaryFiles` from `inputBinding` up to parameter definition" is now in draft-3.  the remaining logic around secondaryFiles is deferred to draft-4.
